### PR TITLE
support wait_for_service

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#215

### CODE REVIEW

File: rmw_opensplice_cpp/src/rmw_service_server_is_available.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include "rmw/allocators.h"
+#include "rmw/error_handling.h"
+#include "rmw/impl/cpp/macros.hpp"
+#include "rmw/rmw.h"
+
+#include "identifier.hpp"
+#include "types.hpp"
+
+// The extern "C" here enforces that overloading is not used.
+extern "C"
```
Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_service_server_is_available.cpp]: do you need to wrap this with `#if __cplusplus` here or is that only needed in header files?

Comment (jacquelinekay) [inline rmw_opensplice_cpp/src/rmw_service_server_is_available.cpp]: Yeah nevermind, I don't think it's needed.

**Comment by wjwwood** [comment]
This is also ready for review. The latest commit 6f65a4e fixed the bug I was struggling with where the publisher and subscription count was always going up and never down.

File: rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/requester.hpp [hunk 2]
```diff
@@ -282,6 +285,46 @@ class Requester
...
+  /// Return NULL on success, otherwise an error string.
+  const char *
+  server_is_available(
+    const rmw_node_t * node, bool * is_available,
+    rmw_ret_t (* count_publishers)(const rmw_node_t *, const char *, size_t *),
+    rmw_ret_t (* count_subscribers)(const rmw_node_t *, const char *, size_t *)
+  ) noexcept
+  {
+    if (!is_available) {
+      return "argument is_available is null";
```
Comment (tfoote) [inline rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/requester.hpp]: should be return nullptr with setting error message?

Comment (wjwwood) [inline rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/requester.hpp]: No, these functions are strange like that. See: https://github.com/ros2/rmw_opensplice/blob/master/rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/service_type_support.h#L62-L66

**Comment by tfoote** [comment]
lgtm, there are several error catches with fprintfs such that a user cannot deal with them. Most are clean up errors. Presumeably if we cannot deallocate things there's something bad happening.  I don't think we have a way to expose that sort of failure at the moment though.

**Comment by wjwwood** [comment]
Yeah, I try to only use `fprintf` when returning an error is not possible, like in the "fail" goto label of a create function.

**Pull request merged**